src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +30,109 @@ public static class Renamer
+
+            if (shouldRenameType)
+            {
+                solution = await RenameTypeToMatchDocumentNameAsync(document.Id, solution, newName, document.Name, cancellationToken).ConfigureAwait(false);
+            }
+
+            if (shouldMoveNamespace)
+            {
+                solution = await SyncNamespaceToDocumentFoldersAsync(document.Id, solution, cancellationToken).ConfigureAwait(false);
+            }
+
+            return new RenameDocumentResult(solution, renameSymbolSucceeded: true);
+        }
+
+        private static async Task<Solution> RenameTypeToMatchDocumentNameAsync(DocumentId documentId, Solution solution, string newName, string originalName, CancellationToken cancellationToken)
+        {
+            var newNameWithoutExtension = Path.GetFileNameWithoutExtension(newName);
+            var oldNameWithoutExtension = Path.GetFileNameWithoutExtension(originalName);
+
+            var document = solution.GetDocument(documentId);
+
+            var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            var typeDeclarations = syntaxRoot.DescendantNodesAndSelf().Where(syntaxFacts.IsTypeDeclaration);
+            var displayNames = typeDeclarations.Select(d => syntaxFacts.GetDisplayName(d, DisplayNameOptions.None));
+
+            var matchingTypeDeclaration = typeDeclarations.FirstOrDefault(d => syntaxFacts.GetDisplayName(d, DisplayNameOptions.None).Equals(oldNameWithoutExtension, StringComparison.OrdinalIgnoreCase));
+            if (matchingTypeDeclaration is object)
+            {
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var symbol = semanticModel.GetDeclaredSymbol(matchingTypeDeclaration, cancellationToken);
+
+                solution = await RenameSymbolAsync(solution, symbol, newNameWithoutExtension, solution.Options, cancellationToken).ConfigureAwait(false);
+            }
+
+            return solution;
+        }
+
+        private static async Task<Solution> SyncNamespaceToDocumentFoldersAsync(DocumentId documentId, Solution solution, CancellationToken cancellationToken)
+        {
+            var document = solution.GetDocument(documentId);
+            var changeNamespaceService = document.GetRequiredLanguageService<IChangeNamespaceService>();
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var targetNamespace = BuildNamespaceFromFolders(document.Folders, syntaxFacts);
+
+            var syntaxRoot = await document.GetSyntaxRootAsync().ConfigureAwait(false);
+
+            var namespaceDeclarations = syntaxRoot.DescendantNodes(descendIntoChildren: n => !syntaxFacts.IsDeclaration(n)).Where(n => !syntaxFacts.IsTypeDeclaration(n));
+
+            foreach (var namespaceDeclaration in namespaceDeclarations)
```

Comment:
1. we're changing *all* the namespaces in the file?
2. it feels like syncing the namespae is potentially something a user would want to opt-into, separate from teh request to rename the type.
3. have we considered not exposing 'RenameDocument' and instead just exposing the individual methods to sync the types/namespace in the file with the Document structure?  user then just does the Doc rename themselves, and then calls these indepedent methods.

Reply:
1. Not all namespaces, only top level 
2. The thought is that this API is surfaced as a "please do things for me", and as a policy we want to sync namespace if possible. Maybe the `OptionSet` would be useful here? to help control behavior? 
3. For now I still prefer a single method. If we need configuration, `OptionSet` seems like a good candidate to help control policy

Reply:
Another idea: the Result APIs returns a IEnumerable<RenameDocumentActions> (RenameDocumentActions is an enum) of things it detected it can do. Then there's a second method you can call that takes the IEnumerable where you say what actual actions you want it to do. If it's list a list then the project system can just take that list and splat it into a WPF checkbox list so if we go add more things we're not having to rev a bunch of UIs.

Reply:
Resolving this as covered in the new proposed design. Start a new thread if concerns still exist.